Sources/SpeziHealthKit/SampleTypeScopedLocalStorage.swift

```diff
@@ -0,0 +1,103 @@
+//
+// This source file is part of the Stanford Spezi open-source project
+//
+// SPDX-FileCopyrightText: 2025 Stanford University and the project authors (see CONTRIBUTORS.md)
+//
+// SPDX-License-Identifier: MIT
+//
+
+
+import Foundation
+import SpeziLocalStorage
+
+
+// Essentially just a thread-safe dictionary that keeps track of our `LocalStorageKey`s used by the `SampleTypeScopedLocalStorage`.
+// The reason this exists is bc the LocalStorage API is intended to be used with long-lived LocalStorageKey objects, which doesn't easily
+// work with the multi-key scoping approach we're using here.
+// Were we not to use something like this for caching and re-using the keys, we'd need to create temporary `LocalStorageKey`s for
+// every load/store operation, which would of course work but would also defeat the whole purpose of having the `LocalStorageKey`s
+// be long-lived objects which are also used for e.g. locking / properly handling concurrent reads or writes.
+private enum LocalStorageKeysHandling {
+    private struct DictKey: Hashable {
+        let valueType: String
+        let sampleType: String
+        
+        init(valueType: (some Any).Type, sampleType: SampleType<some Any>) {
+            // this is fine bc we're not using it as a stable identifier
+            // (the `valueType` key must only be valid&unique for the lifetime of the app)
+            self.valueType = String(reflecting: valueType)
+            self.sampleType = sampleType.id
+        }
+    }
+    
+    private static let lock = NSLock()
```

Comment:
We might use the new `Synchonization` framework here? But I'm also totally fine with keeping `NSLock`, was just a hint if that new framework might be helpful here.

Reply:
the package currently still supports iOS 17, but the `Synchonization` is iOS 18+. if that weren't the case, i'd agree with you that we should use that instead.

Reply:
Makes sense, let's keep `NSLock` then